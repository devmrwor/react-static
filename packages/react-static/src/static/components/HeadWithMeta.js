import React from 'react'
//
import { pathJoin, makePathAbsolute } from '../../utils'
import plugins from '../plugins'
// import packagejson from '../../../package.json'

// const { version } = packagejson

const REGEX_FOR_STYLE_TAG = /<style>|<\/style>/gi

export const InlineStyle = ({ clientCss }) => (
  <style
    key="clientCss"
    type="text/css"
    dangerouslySetInnerHTML={{
      __html: clientCss.toString().replace(REGEX_FOR_STYLE_TAG, ''),
    }}
  />
)

export async function makeHeadWithMeta(state) {
  const {
    head,
    route,
    clientScripts,
    config,
    clientStyleSheets,
    clientCss,
  } = state

  const pluginHeads = await plugins.headElements([], state)

  return ({ children, ...rest }) => {
    const renderLinkCSS = !route.redirect && !config.inlineCss

    const useHelmetTitle =
      head.title && head.title[0] && head.title[0].props.children !== ''

    let childrenArray = React.Children.toArray(children)

    if (useHelmetTitle) {
      head.title[0] = React.cloneElement(head.title[0], { key: 'title' })
      childrenArray = childrenArray.filter(child => {
        if (child.type === 'title') {
          // Filter out the title of the Document in static.config.js
          // if there is a helmet title on this route
          return false
        }
        return true
      })
    }

    const childrenCSS = childrenArray.filter(child => {
      if (
        child.type === 'link' &&
        child.props &&
        child.props.rel === 'stylesheet'
      ) {
        return true
      } else if (child.type === 'style') {
        return true
      }
      return false
    })

    const childrenJS = childrenArray.filter(child => child.type === 'script')
    childrenArray = childrenArray.filter(child => {
      if (
        child.type === 'link' &&
        child.props &&
        child.props.rel === 'stylesheet'
      ) {
        return false
      } else if (child.type === 'style') {
        return false
      } else if (child.type === 'script') {
        return false
      }
      return true
    })

    return (
      <head {...rest}>
        <meta name="generator" content="React Static" />
        {head.base}
        {useHelmetTitle && head.title}
        {head.meta}
        {childrenJS}
        {!route.redirect &&
          clientScripts.map(script => (
            <link
              key={`clientScript_${script}`}
              rel="preload"
              as="script"
              href={makePathAbsolute(
                pathJoin(process.env.REACT_STATIC_ASSETS_PATH, script)
              )}
            />
          ))}
        {childrenCSS}
        {renderLinkCSS &&
          clientStyleSheets.reduce((memo, styleSheet) => {
            const href = makePathAbsolute(
              pathJoin(process.env.REACT_STATIC_ASSETS_PATH, styleSheet)
            )

            return [
              ...memo,
              <link
                key={`clientStyleSheetPreload_${styleSheet}`}
                rel="preload"
                as="style"
                href={href}
              />,
              <link
                key={`clientStyleSheet_${styleSheet}`}
                rel="stylesheet"
                href={href}
              />,
            ]
          }, [])}
        {head.link}
        {head.noscript}
        {head.script}
        {config.inlineCss && <InlineStyle clientCss={clientCss} />}
        {head.style}
        {pluginHeads}
        {childrenArray}
      </head>
    )
  }
}
