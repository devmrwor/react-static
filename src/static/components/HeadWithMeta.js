import React from 'react'

const REGEX_FOR_STYLE_TAG = /<style>|<\/style>/gi

export const InlineStyle = ({ clientCss }) => (
  <style
    key="clientCss"
    type="text/css"
    dangerouslySetInnerHTML={{
      __html: clientCss.toString().replace(REGEX_FOR_STYLE_TAG, ''),
    }}
  />
)

export const makeHeadWithMeta = ({
  head,
  route,
  clientScripts,
  config,
  clientStyleSheets,
  clientCss,
  meta,
}) => ({ children, ...rest }) => {
  const renderLinkCSS = !route.redirect && !config.inlineCss
  const useHelmetTitle = head.title && head.title[0] && head.title[0].props.children !== ''
  let childrenArray = children
  if (useHelmetTitle) {
    head.title[0] = React.cloneElement(head.title[0], { key: 'title' })
    childrenArray = React.Children.toArray(children).filter(child => {
      if (child.type === 'title') {
        // Filter out the title of the Document in static.config.js
        // if there is a helmet title on this route
        return false
      }
      return true
    })
  }

  const pluginHeads = (config.plugins || [])
    .map(plugin => plugin.Head)
    .filter(Boolean)
    .map(PluginHead => <PluginHead meta={meta} />)

  return (
    <head {...rest}>
      {head.base}
      {useHelmetTitle && head.title}
      {head.meta}
      {!route.redirect &&
        clientScripts.map(script => (
          <link
            key={`clientScript_${script}`}
            rel="preload"
            as="script"
            href={`${process.env.REACT_STATIC_PUBLICPATH}${script}`}
          />
        ))}
      {renderLinkCSS &&
        clientStyleSheets.reduce(
          (memo, styleSheet) => [
            ...memo,
            <link
              key={`clientStyleSheetPreload_${styleSheet}`}
              rel="preload"
              as="style"
              href={`${process.env.REACT_STATIC_PUBLICPATH}${styleSheet}`}
            />,
            <link
              key={`clientStyleSheet_${styleSheet}`}
              rel="stylesheet"
              href={`${process.env.REACT_STATIC_PUBLICPATH}${styleSheet}`}
            />,
          ],
          []
        )}
      {head.link}
      {head.noscript}
      {head.script}
      {config.inlineCss && <InlineStyle clientCss={clientCss} />}
      {head.style}
      {pluginHeads}
      {childrenArray}
    </head>
  )
}
