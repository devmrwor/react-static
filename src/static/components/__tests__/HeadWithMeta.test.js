import React from 'react'
import { mount } from 'enzyme'
import { makeHeadWithMeta } from '../HeadWithMeta'

describe('HeadWithMeta', () => {
  let data
  beforeEach(() => {
    process.env.REACT_STATIC_ASSETS_PATH = 'assets/path'
    const inlineCSS = `
      body {
        display: block;
        width: 100%;
        height: 100%;
      }
    `
    data = {
      head: {
        base: '',
        title: [<title>Helmet Title</title>],
        meta: '',
        link: '',
        noscript: '',
        script: '',
        style: '',
      },
      route: { redirect: false },
      embeddedRouteInfo: {
        routeDate: 'here',
      },
      clientScripts: ['main.js', 'bootstrap.js'],
      clientStyleSheets: ['main.css', 'bootstrap.css'],
      clientCss: <style>{inlineCSS}</style>,
      config: {
        inlineCss: false,
      },
    }
  })

  test('when route is a static route', () => {
    const HeadWithMeta = makeHeadWithMeta(data)

    const headWithMeta = mount(
      <HeadWithMeta className="body">
        <meta name="description" content="Helmet application" />
      </HeadWithMeta>
    )

    expect(headWithMeta).toMatchSnapshot()
  })

  test('when route is a redirect route', () => {
    data.route.redirect = true
    const HeadWithMeta = makeHeadWithMeta(data)

    const headWithMeta = mount(
      <HeadWithMeta className="body">
        <meta name="description" content="Helmet application" />
      </HeadWithMeta>
    )

    expect(headWithMeta).toMatchSnapshot()
  })

  test('when route has inline CSS', () => {
    data.config.inlineCss = true
    const HeadWithMeta = makeHeadWithMeta(data)

    const headWithMeta = mount(
      <HeadWithMeta className="body">
        <meta name="description" content="Helmet application" />
      </HeadWithMeta>
    )

    expect(headWithMeta).toMatchSnapshot()
  })

  test('when route has title as child', () => {
    const HeadWithMeta = makeHeadWithMeta(data)

    const headWithMeta = mount(
      <HeadWithMeta className="body">
        <title>Document Title</title>
        <meta name="description" content="Helmet application" />
      </HeadWithMeta>
    )

    expect(headWithMeta).toMatchSnapshot()
  })

  test('when route has no helmet title', () => {
    const HeadWithMeta = makeHeadWithMeta({
      ...data,
      head: {
        ...data.head,
        title: [],
      },
    })

    const headWithMeta = mount(
      <HeadWithMeta className="body">
        <title>Document Title</title>
        <meta name="description" content="Helmet application" />
      </HeadWithMeta>
    )

    expect(headWithMeta).toMatchSnapshot()
  })
})
