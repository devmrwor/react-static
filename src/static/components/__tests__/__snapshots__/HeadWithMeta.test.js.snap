// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadWithMeta when route has inline CSS 1`] = `
<Component
  className="body"
>
  <head
    className="body"
  >
    <title
      key="title"
    >
      Helmet Title
    </title>
    <link
      as="script"
      href="public/pathmain.js"
      key="clientScript_main.js"
      rel="preload"
    />
    <link
      as="script"
      href="public/pathbootstrap.js"
      key="clientScript_bootstrap.js"
      rel="preload"
    />
    <InlineStyle
      clientCss={
        <style>
          
      body {
        display: block;
        width: 100%;
        height: 100%;
      }
    
        </style>
      }
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "[object Object]",
          }
        }
        key="clientCss"
        type="text/css"
      />
    </InlineStyle>
    <meta
      content="Helmet application"
      key=".0"
      name="description"
    />
  </head>
</Component>
`;

exports[`HeadWithMeta when route has no helmet title 1`] = `
<Component
  className="body"
>
  <head
    className="body"
  >
    <link
      as="script"
      href="public/pathmain.js"
      key="clientScript_main.js"
      rel="preload"
    />
    <link
      as="script"
      href="public/pathbootstrap.js"
      key="clientScript_bootstrap.js"
      rel="preload"
    />
    <link
      as="style"
      href="public/pathmain.css"
      key="clientStyleSheetPreload_main.css"
      rel="preload"
    />
    <link
      href="public/pathmain.css"
      key="clientStyleSheet_main.css"
      rel="stylesheet"
    />
    <link
      as="style"
      href="public/pathbootstrap.css"
      key="clientStyleSheetPreload_bootstrap.css"
      rel="preload"
    />
    <link
      href="public/pathbootstrap.css"
      key="clientStyleSheet_bootstrap.css"
      rel="stylesheet"
    />
    <title>
      Document Title
    </title>
    <meta
      content="Helmet application"
      name="description"
    />
  </head>
</Component>
`;

exports[`HeadWithMeta when route has title as child 1`] = `
<Component
  className="body"
>
  <head
    className="body"
  >
    <title
      key="title"
    >
      Helmet Title
    </title>
    <link
      as="script"
      href="public/pathmain.js"
      key="clientScript_main.js"
      rel="preload"
    />
    <link
      as="script"
      href="public/pathbootstrap.js"
      key="clientScript_bootstrap.js"
      rel="preload"
    />
    <link
      as="style"
      href="public/pathmain.css"
      key="clientStyleSheetPreload_main.css"
      rel="preload"
    />
    <link
      href="public/pathmain.css"
      key="clientStyleSheet_main.css"
      rel="stylesheet"
    />
    <link
      as="style"
      href="public/pathbootstrap.css"
      key="clientStyleSheetPreload_bootstrap.css"
      rel="preload"
    />
    <link
      href="public/pathbootstrap.css"
      key="clientStyleSheet_bootstrap.css"
      rel="stylesheet"
    />
    <meta
      content="Helmet application"
      key=".1"
      name="description"
    />
  </head>
</Component>
`;

exports[`HeadWithMeta when route is a redirect route 1`] = `
<Component
  className="body"
>
  <head
    className="body"
  >
    <title
      key="title"
    >
      Helmet Title
    </title>
    <meta
      content="Helmet application"
      key=".0"
      name="description"
    />
  </head>
</Component>
`;

exports[`HeadWithMeta when route is a static route 1`] = `
<Component
  className="body"
>
  <head
    className="body"
  >
    <title
      key="title"
    >
      Helmet Title
    </title>
    <link
      as="script"
      href="public/pathmain.js"
      key="clientScript_main.js"
      rel="preload"
    />
    <link
      as="script"
      href="public/pathbootstrap.js"
      key="clientScript_bootstrap.js"
      rel="preload"
    />
    <link
      as="style"
      href="public/pathmain.css"
      key="clientStyleSheetPreload_main.css"
      rel="preload"
    />
    <link
      href="public/pathmain.css"
      key="clientStyleSheet_main.css"
      rel="stylesheet"
    />
    <link
      as="style"
      href="public/pathbootstrap.css"
      key="clientStyleSheetPreload_bootstrap.css"
      rel="preload"
    />
    <link
      href="public/pathbootstrap.css"
      key="clientStyleSheet_bootstrap.css"
      rel="stylesheet"
    />
    <meta
      content="Helmet application"
      key=".0"
      name="description"
    />
  </head>
</Component>
`;
